Source/Libraries/openXDA.Adapters/HIDSController.cs

```diff
@@ -162,6 +162,78 @@ void BuildQuery(IQueryBuilder builder)
             return response;
         }
 
+        [HttpPost]
+        [ValidateRequestVerificationToken]
+        public async Task<HttpResponseMessage> QueryPointsByTimeSpans([FromBody] JObject query, CancellationToken cancellationToken)
+        {
+            if (!query.ContainsKey("TimeSpans"))
+            {
+                BadRequestErrorMessageResult result = BadRequest("Missing query parameter: TimeSpans");
+                return await result.ExecuteAsync(cancellationToken);
+            }
+
+            IEnumerable<GSF.Range<DateTime>> spans = query["TimeSpans"].ToObject<IEnumerable<JToken>>().Select((spanToken) =>
+            {
+                JArray bounds = (JArray)spanToken;
+
+                return new GSF.Range<DateTime>(bounds[0].Value<DateTime>(), bounds[1].Value<DateTime>());
+            });
+            spans = GSF.Range<DateTime>.MergeAllOverlapping(spans).OrderBy(span => span.Start);
+            if (spans.Count() == 0)
+            {
+                BadRequestErrorMessageResult result = BadRequest("No timespans specified in query parameter");
+                return await result.ExecuteAsync(cancellationToken);
+            }
+
+            IAsyncEnumerable<Point> hidsCaller(API instance, CancellationToken token = default)
+            {
+                IEnumerable<string> tags = Enumerable.Empty<string>();
+                if (query["Channels"] is JArray channels)
+                {
+                    tags = channels
+                        .ToObject<List<int>>()
+                        .Select(channelID => APIExtensions.ToTag(null, channelID));
+                }
+
+                uint invalidFlags = 0;
+                if (query.ContainsKey("InvalidFlags"))
+                {
+                    invalidFlags = query.Value<uint>("InvalidFlags");
+                }
+
+                string aggregateDuration = "";
+                if (query.ContainsKey("AggregateDuration"))
+                {
+                    aggregateDuration = query.Value<string>("AggregateDuration");
+                }
+
+                return spans
+                    .ToAsyncEnumerable()
+                    .SelectMany(span =>
+                    {
+                        void BuildQuery(IQueryBuilder builder)
+                        {
+                            builder.Range(span.Start, span.End);
+                            if (tags.Count() != 0) builder.FilterTags(tags);
```

Comment:
Linq's `Count()` function has to enumerate all the items in the list to determine the final count. The `Any()` function can stop as soon as it has enumerated the first item so it is generally preferred for performance reasons.
```suggestion
                            if (tags.Any()) builder.FilterTags(tags);
```